### PR TITLE
Implemented TestKnnGraph unit tests

### PR DESCRIPTION
**Pull request by msokolov**
Hi @mocobeta this is a set of changes based on your branch. It mostly just adds a unit test and fixes some edge case bugs that came up when running that test. For example it now handles segments with 0 or 1 vector. I also included an alternative merge implementation as part of Lucene90KnnGraphWriter, but it is disabled. I had previously tested it and it works - I think we should start running some performance evaluations now! One concern I have is the number of Java Collections we use, and the copying we do, but let's not optimize until we have done some measurements.  I'll add some inline comments explaining the reasons for the changes

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java [hunk 1]
```diff
@@ -261,14 +262,14 @@ public long ramBytesUsed() {
     final IndexInput dataIn;
 
     int doc = -1;
-    byte[] binaryValue;
+    BytesRef binaryValue;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java]: I am thinking this might later help avoid allocating an object since BytesRef allows for writing into a larger array

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java [hunk 2]
```diff
@@ -382,7 +388,8 @@ public int nextDoc() throws IOException {
 
     @Override
     public int advance(int target) throws IOException {
-      if (doc > target || target > maxDoc) {
+      // enabled random access. Also I think maxDoc really means maxDoc + 1
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java]: the comment should be better -- basically state.segmentInfo.maxDoc() really is the *number* of documents in the segment - it's a historical misnaming in Lucene. Maybe we could use a better naming in this class?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java [hunk 3]
```diff
@@ -397,7 +404,7 @@ public int advance(int target) throws IOException {
       }
       dataIn.seek(offset);
       int maxLevel = dataIn.readInt();
-      assert maxLevel >= 0;
+      assert maxLevel >= 0 || entry.enterPoints.length == 0;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java]: oh this should be reverted - I toyed with having maxLevel = -1 for single-node graphs, but later reverted that

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java [hunk 4]
```diff
@@ -407,17 +414,19 @@ private FriendsRef readFriends(int maxLevel) throws IOException {
...
-        assert friendSize > 0;
-        int[] frineds = new int[friendSize];
-        int friendId = dataIn.readVInt();  // first friend id
-        frineds[0] = friendId;
-        for (int i = 1; i < friendSize; i++) {
-          int delta = dataIn.readVInt();
-          assert delta > 0;
-          friendId += delta;
-          frineds[i] = friendId;
+        // assert friendSize > 0;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphReader.java]: we might have a one-node graph

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java [hunk 5]
```diff
@@ -76,7 +84,10 @@ public Lucene90KnnGraphWriter(SegmentWriteState state) throws IOException {
 
   @Override
   public void writeField(FieldInfo fieldInfo, KnnGraphReader values) throws IOException {
-    Map<Integer, Long> docToOffset = new HashMap<>();
+    // we require a TreeMap so that we can iterate over docid keys in order. Instead we should move
+    // to the IndexedDISI encoding used by BinaryDocValues
+    Map<Integer, Long> docToOffset = new TreeMap<>();
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java]: The ordering really matters here! Otherwise the vectors and the neighbors can get out of sync

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java [hunk 6]
```diff
@@ -103,28 +112,39 @@ public void writeField(FieldInfo fieldInfo, KnnGraphReader values) throws IOExce
...
-        // sort friend ids
-        Arrays.sort(friends.ints, friends.offset, stop);
-        // write the smallest friend id
-        graphData.writeVInt(friends.ints[friends.offset]);
-        for (int i = friends.offset + 1; i < stop; i++) {
-          // write delta
-          int delta = friends.ints[i] - friends.ints[i - 1];
-          assert delta > 0;
-          graphData.writeVInt(delta);
+        if (friends.length > 0) {
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java]: again handling 1-member graph

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java [hunk 7]
```diff
@@ -103,28 +112,39 @@ public void writeField(FieldInfo fieldInfo, KnnGraphReader values) throws IOExce
...
     long vectorDataLength = vectorData.getFilePointer() - vectorDataOffset;
     long graphDataLength = graphData.getFilePointer() - graphDataOffset;
 
     writeMeta(fieldInfo, vectorDataOffset, vectorDataLength, graphDataOffset, graphDataLength, graph.getTopLevel(), graph.getEnterPoints(), docToOffset);
   }
 
+  private void writeVectorValue(int numDims, VectorValues vectors) throws IOException {
+    // write vector value
+    BytesRef binaryValue = vectors.binaryValue();
+    VectorValues.verifyNumDimensions(binaryValue.length, numDims);
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java]: we probably should not be doing this verification for every value - haven't we already checked this when the VectorField was added?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java [hunk 8]
```diff
@@ -139,11 +159,197 @@ private void writeMeta(FieldInfo field, long vectorDataOffset, long vectorDataLe
     }
     meta.writeInt(docToOffset.size());
     for (Map.Entry<Integer, Long> entry : docToOffset.entrySet()) {
+      // these are not in sorted order, yet we write the vectors in order by docid
       meta.writeVInt(entry.getKey());
       meta.writeVLong(entry.getValue());
     }
   }
 
+  /*
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90KnnGraphWriter.java]: If we uncomment this, we can try the simple merging by re-inserting every element.

File: lucene/core/src/java/org/apache/lucene/index/SegmentMerger.java [hunk 9]
```diff
@@ -100,15 +101,8 @@ MergeState merge() throws IOException {
       throw new IllegalStateException("Merge would result in 0 document segment");
     }
     mergeFieldInfos();
-    long t0 = 0;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/index/SegmentMerger.java]: Most of this is just unrelated refactoring since I found it painful to again duplicate all this repeated logging boilerplate.

File: lucene/core/src/java/org/apache/lucene/index/VectorValues.java [hunk 10]
```diff
@@ -78,33 +81,28 @@ public static float distance(float[] v1, float[] v2, DistanceFunction distFunc)
...
+    verifyNumDimensions(bytes.length, numDims);
     float[] value = new float[numDims];
-    for (int i = 0; i < numDims; i++) {
-      int bits = ((bytes[i * Float.BYTES] & 0xFF) << 24) |
-          ((bytes[i * Float.BYTES + 1] & 0xFF) << 16) |
-          ((bytes[i * Float.BYTES + 2] & 0xFF) << 8) |
-          (bytes[i * Float.BYTES + 3] & 0xFF);
-      value[i] = Float.intBitsToFloat(bits);
-    }
+    ByteBuffer buffer = ByteBuffer.wrap(bytes.bytes, bytes.offset, bytes.length);
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/index/VectorValues.java]: Let's use the ByteBuffer/FloatBuffer to do these - they are probably tuned assembly routines that are faster

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java [hunk 11]
```diff
@@ -37,14 +37,13 @@
 public final class HNSWGraph implements Accountable {
 
   private final VectorValues.DistanceFunction distFunc;
-  private final Map<Integer, Layer> layers;
+  private final List<Layer> layers;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java]: the map keys were redundant since we have the layer number in the Layer and it is implicit in the list position

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java [hunk 12]
```diff
@@ -37,14 +37,13 @@
...
 
-  private boolean freeze = false;
-  private int topLevel = 0;
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java]: topLevel seems to be == layers.size() - 1?

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java [hunk 13]
```diff
@@ -100,12 +97,12 @@ public NearestNeighbors searchLayer(float[] query, Neighbor ep, int ef, int leve
         }
         float dist = distance(query, e.docId(), vectorValues);
         if (dist < f.distance() || results.size() < ef) {
-          if (results.size() == ef) {
-            results.pop();
-          }
           Neighbor n = new ImmutableNeighbor(e.docId(), dist);
           candidates.insertWithOverflow(n);
-          results.add(n);
+          Neighbor popped = results.insertWithOverflow(n);
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HNSWGraph.java]: I think it is more efficient to insertWithOverflow than to pop when full. There are other efficiency changes we can make here, but let's come back to that later.

File: lucene/core/src/test/org/apache/lucene/index/TestKnnGraph.java [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+import static org.apache.lucene.util.hnsw.HNSWGraphWriter.RAND_SEED;
+
+/** Tests indexing of a knn-graph */
+public class TestKnnGraph extends LuceneTestCase {
+
+  private static final String KNN_GRAPH_FIELD = "vector";
+
+  @Before
+  public void setup() {
+    RAND_SEED = random().nextLong();
```
Comment (msokolov) [inline lucene/core/src/test/org/apache/lucene/index/TestKnnGraph.java]: Makes randomized tests repeatable

**Comment by mocobeta** [comment]
Thanks @msokolov I am fine with the changes. 
I'm a bit busy for a while, please feel free to directly merge commits into this repo without waiting for me to make further improvements (you have write permission here). If you need my comments I'd be happy to review the PRs but my response might be delayed...

> I think we should start running some performance evaluations now!
Agreed. Especially repeated calls to searchLayer() method are very time consuming, can we optimize this?
